File: EstherCharLib/fetch-cgtw-names.mjs
/**
 * 从 CGTeamwork (proj_eth) 查询 Cha 类型资产的中文名和缩略图
 * 输出: cgtw-asset-names.json + thumbnails/ 文件夹
 */

import fs from "node:fs";
import path from "node:path";
import { pipeline } from "node:stream/promises";
import { Readable } from "node:stream";

process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";

const CGTW_HOST = "cgt-artstudio.xindong.com";
const BASE_URL = `https://${CGTW_HOST}`;
const API_URL = `${BASE_URL}/api.php`;
const ACCOUNT = "wuyixiang@xd.com";
const PASSWORD = "ares1121qqwq";

const SCRIPT_DIR = path.dirname(new URL(import.meta.url).pathname.replace(/^\/([A-Z]:)/, "$1"));
const THUMB_DIR = path.join(SCRIPT_DIR, "thumbnails");

async function apiCall(controller, method, data = {}, token = "") {
  const payload = { controller, method, app: "api", ...data };
  const formBody = "data=" + encodeURIComponent(JSON.stringify(payload));
  const headers = { "Content-Type": "application/x-www-form-urlencoded" };
  if (token) headers["Cookie"] = `token=${token}`;
  const resp = await fetch(API_URL, { method: "POST", headers, body: formBody });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
  const result = await resp.json();
  if (result && result.code !== "1") {
    throw new Error(`CGTW error: ${JSON.stringify(result.data || result)}`);
  }
  return result?.data ?? result;
}

async function login() {
  console.log(`登录 ${CGTW_HOST} ...`);
  const data = await apiCall("account", "login", { account: ACCOUNT, password: PASSWORD, machine_key: "" });
  if (!data?.token) throw new Error("登录失败");
  console.log(`登录成功`);
  return data.token;
}

async function downloadImage(url, destPath, token) {
  try {
    const headers = {};
    if (token) headers["Cookie"] = `token=${token}`;
    const resp = await fetch(url, { headers });
    if (!resp.ok || !resp.body) return false;
    await pipeline(Readable.fromWeb(resp.body), fs.createWriteStream(destPath));
    return true;
  } catch { return false; }
}

async function main() {
  const token = await login();

  const DB = "proj_film";
  const MOD = "m_asset";

  console.log(`查询 ${DB}.${MOD} (type=Cha) ...`);
  const assets = await apiCall("info", "get_filter", {
    db: DB, module: MOD,
    sign_array: [`${MOD}.entity`, `${MOD}.cn_name`, `${MOD}.type`, `${MOD}.image`, `${MOD}.notes`, `${MOD}.status`, `${MOD}.first_appear`, `${MOD}.create_time`],
    sign_filter_array: [[`${MOD}.type`, "=", "Cha"]],
    order_sign_array: [`${MOD}.entity`],
    limit: "5000", start_num: "",
  }, token);

  if (!Array.isArray(assets) || !assets.length) {
    console.log("未查到 Cha 资产"); return;
  }
  console.log(`共 ${assets.length} 个 Cha 资产`);

  if (!fs.existsSync(THUMB_DIR)) fs.mkdirSync(THUMB_DIR, { recursive: true });

  const result = [];
  let dlCount = 0;
  for (const a of assets) {
    const entity = a[`${MOD}.entity`] || "";
    const cnName = a[`${MOD}.cn_name`] || "";
    const notesRaw = (a[`${MOD}.notes`] || "").trim();
    const imageRaw = a[`${MOD}.image`] || "";

    const notesParts = notesRaw.split("|").map(s => s.trim());
    const status = (a[`${MOD}.status`] || "").trim();
    const firstAppear = (a[`${MOD}.first_appear`] || "").trim();
    const createTime = (a[`${MOD}.create_time`] || "").trim();
    const entry = {
      entity,
      cn_name: cnName,
      status,
      first_appear: firstAppear,
      create_time: createTime,
      notes: notesRaw,
      notes_episode: notesParts[0] || "",
      notes_cn: notesParts[1] || "",
      notes_en: notesParts[2] || "",
      notes_role: notesParts[3] || "",
      thumbnail: "",
    };

    let imageUrl = "";
    if (imageRaw) {
      try {
        const parsed = JSON.parse(imageRaw);
        if (Array.isArray(parsed) && parsed.length) {
          imageUrl = parsed[0].max || parsed[0].min || "";
        }
      } catch {
        imageUrl = imageRaw;
      }
    }

    if (imageUrl) {
      if (!imageUrl.startsWith("http")) {
        imageUrl = `${BASE_URL}${imageUrl.startsWith("/") ? "" : "/"}${imageUrl}`;
      }
      const ext = path.extname(new URL(imageUrl).pathname) || ".png";
      const thumbName = `${entity}${ext}`;
      const thumbPath = path.join(THUMB_DIR, thumbName);

      if (fs.existsSync(thumbPath)) {
        entry.thumbnail = thumbName;
      } else {
        const ok = await downloadImage(imageUrl, thumbPath, token);
        if (ok) {
          entry.thumbnail = thumbName;
          dlCount++;
          if (dlCount % 20 === 0) console.log(`  已下载 ${dlCount} 张...`);
        }
      }
    }

    result.push(entry);
  }

  const outPath = path.join(SCRIPT_DIR, "cgtw-asset-names.json");
  fs.writeFileSync(outPath, JSON.stringify(result, null, 2), "utf-8");
  console.log(`\n完成: ${result.length} 个 Cha 资产, 下载 ${dlCount} 张缩略图`);
  console.log(`数据: ${outPath}`);
  console.log(`缩略图: ${THUMB_DIR}`);

  const withThumb = result.filter(r => r.thumbnail);
  console.log(`有缩略图: ${withThumb.length} / ${result.length}`);
}

main().catch(e => { console.error("错误:", e.message); process.exit(1); });

File: TorchCharLib/fetch-cgtw-names.py
"""
从 CGTeamwork 查询 TLI 项目角色中文名
直接使用 requests 调用 CGTW REST API，不依赖 cgtw2 SDK
"""

import json
import requests
import sys
import os

requests.packages.urllib3.disable_warnings()

CGTW_HOST = "cgt-artstudio.xindong.com"
CGTW_PORT = 8443
ACCOUNT = "wuyixiang@xd.com"
PASSWORD = "ares1121qqwq"

BASE_URL = f"https://{CGTW_HOST}:{CGTW_PORT}"
API_URL = f"{BASE_URL}/api.php"

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))

session = requests.Session()
session.verify = False


def api_call(controller, method, data=None, token=""):
    payload = {
        "controller": controller,
        "method": method,
        "token": token,
    }
    if data:
        payload.update(data)
    resp = session.post(API_URL, json=payload, timeout=30)
    resp.raise_for_status()
    result = resp.json()
    if isinstance(result, dict) and result.get("code") == "1":
        raise Exception(f"CGTW API error: {result.get('data', result)}")
    if isinstance(result, dict) and "data" in result:
        return result["data"]
    return result


def login():
    print(f"正在登录 {CGTW_HOST}:{CGTW_PORT} ...")
    data = api_call("account", "login", {
        "account": ACCOUNT,
        "password": PASSWORD,
        "machine_key": "",
    })
    if not isinstance(data, dict) or "token" not in data:
        raise Exception(f"登录失败: {data}")
    print(f"登录成功: {data.get('account', ACCOUNT)}")
    return data["token"]


def get_projects(token):
    print("\n查询所有项目数据库...")
    data = api_call("c_orm", "get_with_filter", {
        "db": "public",
        "module": "project",
        "module_type": "info",
        "sign_array": ["project.entity", "project.database", "project.status", "project.code"],
        "sign_filter_array": [],
        "order_sign_array": [],
        "limit": "200",
        "start_num": "",
    }, token)
    return data


def get_modules(token, db):
    print(f"\n查询 {db} 的模块...")
    data = api_call("info", "modules", {"db": db}, token)
    return data


def get_fields(token, db, module):
    print(f"\n查询 {db}.{module} 的字段...")
    data = api_call("info", "fields_and_str", {"db": db, "module": module}, token)
    return data


def get_assets(token, db, field_list):
    print(f"\n查询 {db} asset 数据...")
    data = api_call("info", "get_filter", {
        "db": db,
        "module": "asset",
        "sign_array": field_list,
        "sign_filter_array": [],
        "order_sign_array": ["asset.entity"],
        "limit": "5000",
        "start_num": "",
    }, token)
    return data


def main():
    token = login()

    # Step 1: 查找 TLI 项目数据库
    projects = get_projects(token)
    if projects:
        print("\n所有项目:")
        tli_db = None
        for p in projects:
            name = p.get("project.entity", "")
            db = p.get("project.database", "")
            code = p.get("project.code", "")
            status = p.get("project.status", "")
            print(f"  {name} | 代号:{code} | 数据库:{db} | 状态:{status}")
            if code and "TLI" in code.upper():
                tli_db = db
            elif name and "TLI" in name.upper():
                tli_db = db
        if tli_db:
            print(f"\n找到 TLI 数据库: {tli_db}")
        else:
            print("\n未找到 TLI 项目，请手动确认数据库名")
            return
    else:
        print("未能获取项目列表，尝试常见数据库名...")
        tli_db = "proj_tli"

    # Step 2: 查看 asset 模块字段
    fields = get_fields(token, tli_db, "asset")
    if fields:
        print(f"\nasset 模块字段 ({len(fields)} 个):")
        for f in fields:
            sign = f.get("sign", "")
            cn = f.get("str", "")
            print(f"  {sign:40s} → {cn}")

    # Step 3: 查询角色数据
    query_fields = ["asset.entity", "asset.cn_name", "asset.asset_name"]
    valid_fields = [qf for qf in query_fields if any(f.get("sign") == qf for f in (fields or []))]
    if not valid_fields:
        valid_fields = ["asset.entity"]
        if fields:
            for f in fields:
                s = f.get("sign", "")
                n = f.get("str", "").lower()
                if "名" in n or "name" in n.lower():
                    valid_fields.append(s)
            valid_fields = list(dict.fromkeys(valid_fields))

    print(f"\n使用字段: {valid_fields}")
    assets = get_assets(token, tli_db, valid_fields)
    if assets:
        print(f"\n角色数据 ({len(assets)} 条):")
        for a in assets:
            print(f"  {json.dumps(a, ensure_ascii=False)}")

        out_path = os.path.join(SCRIPT_DIR, "cgtw-asset-names.json")
        with open(out_path, "w", encoding="utf-8") as fp:
            json.dump(assets, fp, ensure_ascii=False, indent=2)
        print(f"\n已保存到: {out_path}")
    else:
        print("未查询到资产数据")


if __name__ == "__main__":
    main()
